ODIN_II/SRC/include/HardSoftLogicMixer.hpp

```diff
@@ -0,0 +1,105 @@
+#include "netlist_analyzer.h"
+
+class HardSoftLogicMixer{
+public:
+    HardSoftLogicMixer(t_arch& arch,const config_t configuration,std::vector<t_physical_tile_type> tileTypes);
+    /*----------------------------------------------------------------------
+    * Function: calculateGridSizes 
+    *   For all the layouts represented in the architecture file, populates
+    *   the _grid_layout_sizes map with corresponding estimated sizes
+    *---------------------------------------------------------------------*/
+    void calculateAllGridSizes();
+
+    /* Returns whether the hard blocks and soft logic implementation 
+     * of multipliers should be mixed in the process of synthesis
+     */
+    bool mixHardBlocksOfType(HardBlocksOptimizationTypesEnum type);
+    /*----------------------------------------------------------------------
+     * Function: selectLogicToImplementInHardBlocks 
+     *  For all the notes, that noted to be candidates to be implemented in the 
+     * hard blocks, launches corresponding logic
+     * Parameters: 
+     * returns:
+     *---------------------------------------------------------------------*/
+    void selectLogicToImplementInHardBlocks(netlist_t *netlist);
+
+    /*----------------------------------------------------------------------
+     * Function: selectLogicToImplementInHardBlocks 
+     * Calculates number of available hard blocks by issuing a call,
+     * traverses the netlist and statistics to figure out
+     * which operation should be implemented on the hard block
+     * Parameters: 
+     *      node_t * : pointer to candidate node
+     *      HardBlocksOptimizationTypesEnum : type of the hard block to optimize
+     * returns:
+     *---------------------------------------------------------------------*/
+    void takeNoteOfAPotentialHardBlockNode( nnode_t * node, HardBlocksOptimizationTypesEnum type);
+
+    
+private:
+    /* This function parses the configuration bitmask and sets
+     * the corresponding boolean variables in the _enabledOptimizationsArray.
+     */
+    void parseAndSetOptimizationParameters(const config_t);
+    int countHardBlocksInArch(t_grid_def& layout ,int hardBlockType,std::pair<int,int> size);
+    void scaleHardBlockCounts();
+
+    void implementUnassignedLogicInSoftLogic(netlist_t* netlist);
+    int inferHardBlocksFromNetlist(int currentOptimizationKind);
+    void chooseHardBlocks(netlist_t *netlist,HardBlocksOptimizationTypesEnum type);
+    /* This map holds estimated device sizes that would
+     * correspond to the architecture, keyed by the pointers to 
+     * layout  description
+     */
+    std::map<int,std::pair<int,int>> _grid_layout_sizes; 
+
+    // This array is composed of vectors, that store nodes that
+    // are potential candidates for performing mixing optimization
+    std::vector<nnode_t*>potentialHardBlockNodes[HardBlocksOptimizationTypesEnum::Count];
+    int _hardBlocksCount [HardBlocksOptimizationTypesEnum::Count];
```

Comment:
Why the _ before the name? Unless it is to match the naming style of the file (they are mostly use to hide function behind define wrapper in odin_error.h) this makes it confusing

Reply:
It is usually used as a prefix for private variable names as a visual indication that the scope of the variable is private to the class in which it is declared. (from stackoverflow). Shall I change it?